tables/validate.py

```diff
@@ -0,0 +1,132 @@
+
+import pandas as pd
+import pandera as pa
+
+from tables.table import Table
+
+
+@dataclass
+class Failure:
+    """Validation failure information returned by Pandera."""
+
+    schema_context: str
+    column: str
+    check: str
+    check_number: int
+    failure_case: str
+    index: int
+
+
+class ValidationError:
+    """Validation error with reference to an Excel cell."""
+
+    message: str
+    excel_cell_ref: str | None
+
+    def __init__(self, message: str, excel_cell_ref: str | None):
+        self.message = message
+        self.excel_cell_ref = excel_cell_ref
+
+
+class TableValidationError(Exception):
+    validation_errors: list[ValidationError]
+
+    def __init__(self, validation_errors: list[ValidationError]):
+        self.validation_errors = validation_errors
+
+
+class TableValidator:
+    """
+    Validates a table against a specified schema.
+
+    Attributes:
+        schema (pa.DataFrameSchema): The schema to validate against.
+        MESSAGE_OVERRIDE (dict[str, str]): A dictionary mapping validation failure messages
+            to custom override messages.
+        IGNORED_FAILURES (dict[str, str]): A dictionary mapping Failure attributes to regular expressions
+            for ignored validation failures. These regex patterns can be used to exclude
+            certain types of validation checks.
+
+    Example usage:
+    >>> schema_config = {
+    >>>     "columns": {
+    >>>         "name": pa.Column(str, checks=[pa.Check.str_length(max_value=50)]),
+    >>>         "age": pa.Column(int, checks=[pa.Check.greater_than(0)]),
+    >>>     }
+    >>> }
+    >>> validator = TableValidator(schema_config)
+    >>> table = Table(...)  # Assume you have a table instance
+    >>> try:
+    >>>     validator.validate(table)
+    >>> except TableValidationError as e:
+    >>>     for validation_error in e.validation_errors:
+    >>>         print(validation_error)
+    """
+
+    schema: pa.DataFrameSchema
+    MESSAGE_OVERRIDE = {
+        "not_nullable": "The cell is blank but is required.",
+        "field_uniqueness": "You entered duplicate data. Remove or replace the duplicate data.",
+        "multiple_fields_uniqueness": "You entered duplicate data. Remove or replace the duplicate data.",
+    }
+    IGNORED_FAILURES = {
+        "failure_case": r"TypeError\(.*",  # catches check failures on data of the wrong type
+        "check": r"dtype\(.*",  # catches dtype failures
+    }
+
+    def __init__(self, schema_config: dict[str, Any]) -> None:
+        """Initialises a TableValidator.
+
+        :param schema_config: A dictionary containing schema configuration.
+            The schema configuration should include information about columns, data types,
+            and any additional validation rules.
+        :raises ValueError: If the schema has coercion enabled.
+        """
+        if schema_config.get("coerce") or any(column.coerce for column in schema_config.get("columns", {}).values()):
```

Comment:
What is this "Coercion not supported" thing - is this still correct now that we are doing type coercion?

Reply:
I think it would be sensible to support coercion going forwards so that we can coerce the data if all validation (including fund specific) passes, before passing it to transformation. Think this should be done in a subsequent PR though. I was thinking that we may want a different class or possible another public method on `TableValidator` that would carry this out

Reply:
Okay fair enough - my understanding was that the reason we re-introduced the types as first positional arguments to `Column` instances in the schema definition was because we wanted to coerce the data before transformation - is this the case?

Reply:
Yes that's the intention. I began to add coercion back but it opened up more complexity which I'd rather just address in another PR.